### DISCUSSION TITLE
resolution of  Literal Types.Union from const

### DISCUSSION THREAD

**Discussion by ahhmarr** [sinclairzx81/typebox#315] [Q&A] [answered]
I've my const such as 

`const Allowed = ["chips", "banana", "apple"] as const `

need to convert it into 
```
Types.Union([
  Type.Literal("chips"),
  Type.Literal("banana"),
  Type.Literal("apple"),
])
```

is it possible with generic helpers provided by TypeBox ?

**Comment by sinclairzx81** [comment, marked as answer]
@ahhmarr Hi, you can try something like the following

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUDGAFhiAIYDyYMwEAdgM464ynX5wC+cAZlBCHADkAAXrBa+ADalgUAPQx0GAEYQAHoICwAKB0Y1kWDwCuE6nTgBVWjVrluzKOIDmAHmRx9MDLQAm9OHoYJ1pnAG0AXQA+AAoAN1JJYwx6AC44KAxSXzpJVDgwgDpi5AicCCpbNLwiEgpKugCAXkQOAEpEHTgMjBhjKFoUJULrW3jE5PpCsjBxpIw4JqihzEKAGWBvKES55La28oaGNp0OHR18RvgAQUlJCAB3DF9FgoAiImAwejecN+VSLRAaRfnA3qQwGBJBg3hE4KQApcGPBztokUEUK9RnR7I4XDFbvcnr4TrptIpMJiWsxWMB8K4KRgINwUFEgA)

```typescript
import { Type, Static, SchemaOptions } from '@sinclair/typebox'

export function UnionOfString<T extends string[]>(values: readonly [...T], options: SchemaOptions = {}) {
  return Type.Union(values.map(value => Type.Literal(value)), options)
}

const Allowed = ["chips", "banana", "apple"] as const 

const T = UnionOfString(Allowed)

type T = Static<typeof T> // type T = "chips" | "banana" | "apple"
```
Hope this helps
S

**Reply by thes01** [reply]
There's one issue with this approach - it doesn't work with `StaticDecode` as the inferred type is `never`. I consider this being a bug @sinclairzx81. So far, I have been successful with this:

```js
function TLiterals<const T extends string[]>(...literal_values: T) {
	return literal_values.map((v) => Type.Literal(v)) as {
		[Key in keyof T]: TLiteral<T[Key]>;
	};
}
```
and then using `Type.Union(TLiterals('a', 'b', 'c'))`

**Reply by sinclairzx81** [reply]
@thes01 Hi, 

You can try the following revised pattern. This pattern generally recommended for applying .map operations in the type system where the updated type is mapping from `string[]` to `TLiteral[]`. The technique is tail safe recursive, with the accumulator (`Acc`) tending to scale better for larger tuples in the TS type system.

[TypeScript Link Here](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxjvkaQCIYkQAm2eJAFhiAQPJjEQB2AZxzJcrdgWEBVPsH7CAMsBgYoBADZwAvnABmUCCDgByAAIDgfEmoLAoAehjoMAIwgAPIwChPdu3ACuAhhwMCzAAnBgRMpQfCEQcAB0HGAAFACUcAIwUBYA5gDaALpwFjAJyIox6sWeGG6QsCFOKNKyfJw6+Ll8eQA8yHD1ynz0Edk9hUU4AIIkJENuI2MoomwcxXAAvHDFAHzbcJ5wKIvLEQUWOipw8mcYo+M5+TiJb1c3AEr3j1nPvcUiscTnAAPytGT8TrdfJ9T44ApvRJzMgoKoqdR9eR7Ip7YEnABccBR3nqjXgOn8ll4cTaUK6-36dEE8EGwweKwm+X2qQAbup-BgBETEW9kNM4BAeO1hcx1lxpfwIjsEJp0kTkHS+AMtdDGQM9gdEMCoBgYP5YignIktXyBULkgQ0vy1ILtgc0JhEujVGo7a6MOl0jgpTSBJkCBE+BheTdfFkCNdPJpvMzsqcdrqGZNUgUAESsYBgAR5nB55wEPiVgiluB5p1gNQYPMldKeRyYPCHCjEEh9DsYCA6FBGkEndstKjdwi9mh0Rj9pxDkdweMDjN1wvFvNwAA+dYrVarO-39bAjebQA)

```typescript
import { Type, Static, StaticDecode, SchemaOptions, TSchema, TUnion, TLiteral } from '@sinclair/typebox'

// use this pattern to .map() string[] into TLiteral[]
export type TUnionOfString<T extends string[], Acc extends TSchema[] = []> = 
  T extends [infer L extends string, ...infer R extends string[]]
    ? TUnionOfString<R, [...Acc, TLiteral<L>]>
    : Acc

export function UnionOfString<const T extends string[]>(values: [...T], options: SchemaOptions = {}): TUnion<TUnionOfString<T>>  {
  return Type.Union(values.map(value => Type.Literal(value)), options) as never // safe
}

const T = UnionOfString(["chips", "banana", "apple"] )
type S = Static<typeof T>       
type D = StaticDecode<typeof T> // type T = "chips" | "banana" | "apple"
```
I use this technique internal to TypeBox. 

Hope this helps
S

**Reply by thes01** [reply]
Thank you!

**Reply by hunterwilhelm** [reply]
I would love to see this in the documentation!

**Reply by hunterwilhelm** [reply]
Made a npm package that includes `UnionOfString`. I'll add more of these helper functions as I find them, because it doesn't seem that Typebox's scope consists of these utility functions.

`npm i @hunterstack/typebox-helpers`

Src: https://github.com/hunterwilhelm/typebox-helpers
